Do not report CA1812 (Make static) on public APIs for web projects

mavasani opened the pull request:
Fixes #3835

Evangelink commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.cs:
```diff
@@ -308,24 +319,6 @@ void Add(INamedTypeSymbol? symbol)
             return builder?.ToImmutable() ?? ImmutableArray<INamedTypeSymbol>.Empty;
         }
 
-        private static bool IsSkippedMethod(IMethodSymbol methodSymbol, WellKnownTypeProvider wellKnownTypeProvider)
```
Is there any need to keep all the web attribute exclusions (see https://github.com/dotnet/roslyn-analyzers/pull/3875/files#diff-b71c5f4c8942242dec4740ebf6efda72R282-R298) now that we don't report on public APIs for web projects?

mavasani replied:
Hmm @pranavkm for advice. Can these web attributes be applied on non-public methods? Unless we are sure, I'd prefer to keep the attribute based exclusions.

pranavkm replied:
The link is bugged (I can't see where it's meant to go), but in general MVC (5.0 and Core) will only discover public types and methods.

Evangelink replied:
@pranavkm Here is the link on master: https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.cs#L270-L287

pranavkm replied:
Those are required to be on public types or methods. Feel free to remove these exclusions.

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStaticTests.cs:
```diff
@@ -1016,9 +1016,18 @@ public class C : System.Web.HttpApplication
 
-            await VerifyVB.VerifyAnalyzerAsync(@"
+            const string editorConfigText = "build_property.UsingMicrosoftNETSdkWeb = true";
+            var csTest = new VerifyCS.Test()
```
Any reason you have split the test and the `RunAsync`? You could write something like

```csharp
await new VerifyCS.Test
{
	TestCode = csSource,
	AnalyzerConfigDocument = editorConfigText,
}.RunAsync();
```

Evangelink replied:
In case you apply this comment, note that it applies to the various tests you have changed.

mavasani replied:
Awaiting an object creation with initializer with an async method invocation seems too chatty to me :-). I prefer to keep object creation with initializer as its own statement.

mavasani replied:
Seems like other tests in this file follow the pattern you suggested. I will switch the pattern in new tests to make it consistent.

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStaticTests.cs:
```diff
@@ -1016,9 +1016,18 @@ public class C : System.Web.HttpApplication
+            var csTest = new VerifyCS.Test()
+            {
+                TestCode = csSource,
+                AnalyzerConfigDocument = editorConfigText
```
[minor] @sharwell likes to have the trailing comma here :)

```suggestion
                AnalyzerConfigDocument = editorConfigText,
```

Evangelink replied:
If you decide to apply this comment, note that you will have to do the change for the various tests you have changed.

mavasani replied:
Sure, will change.

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStaticTests.cs:
```diff
@@ -1016,9 +1016,18 @@ public class C : System.Web.HttpApplication
+}";
 
-            await VerifyVB.VerifyAnalyzerAsync(@"
+            const string editorConfigText = "build_property.UsingMicrosoftNETSdkWeb = true";
```
Just to confirm I got things right, we don't expect user to set this in their `.editorconfig` file? This is a msbuild property that is present for web apps, right?

mavasani replied:
Yes, users do not have to do anything.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.cs:
```diff
@@ -191,7 +194,15 @@ private static bool ShouldAnalyze(IMethodSymbol methodSymbol, WellKnownTypeProvi
+
+            // Do not analyze public APIs for web projects
+            // See https://github.com/dotnet/roslyn-analyzers/issues/3835 for details.
+            if (isWebProject && methodSymbol.IsExternallyVisible())
```
Shall we really skip `protected` members? I don't know exactly if they expect the member to be `public` for the reflection or if `protected` is ok.

mavasani replied:
I am not sure, but I made is as conservative as possible, given protected members in public types are public APIs.

Evangelink replied:
It seems that `Application_Start` could be `protected` so it's mandatory anyways.

Evangelink commented on src/Utilities/Compiler/Extensions/CompilationExtensions.cs:
```diff
@@ -59,5 +61,15 @@ internal static bool TargetsDotNetFramework(this Compilation compilation)
+        internal static bool IsWebProject(this Compilation compilation, AnalyzerOptions options, CancellationToken cancellationToken)
+        {
+            var usingMicrosoftNETSdkWeb = options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.UsingMicrosoftNETSdkWeb, compilation, cancellationToken);
+            return usingMicrosoftNETSdkWeb != null && bool.TryParse(usingMicrosoftNETSdkWeb, out var result) && result;
```
Do you need to potentially trim the string option before doing a `TryParse`? Also as you are then checking the fact this is true, isn't it better to replace the full line with `string.Equals(usingMicrosoftNETSdkWeb?.Trim(), "true", StringComparison.OrdinalIgnoreCase)`?

Evangelink commented on src/Utilities/Compiler/Extensions/CompilationExtensions.cs:
```diff
@@ -59,5 +61,15 @@ internal static bool TargetsDotNetFramework(this Compilation compilation)
+        /// </summary>
+        internal static bool IsWebProject(this Compilation compilation, AnalyzerOptions options, CancellationToken cancellationToken)
+        {
+            var usingMicrosoftNETSdkWeb = options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.UsingMicrosoftNETSdkWeb, compilation, cancellationToken);
```
[cosmetic] the variable name reads to me as if it's a boolean while it's a string, maybe `usingMicrosoftNETSdkWebPropertyValue` is clearer (not sure because it's really long...).

Evangelink commented on src/Utilities/Compiler/Options/AbstractCategorizedAnalyzerConfigOptions.cs:
```diff
@@ -88,13 +89,23 @@ public bool TryGetOptionValue<T>(string optionName, OptionKind kind, DiagnosticD
+                    TryGetSpecificOptionValue(rule.Category, optionKeyPrefix, out optionValue) ||
+                    TryGetAnySpecificOptionValue(rule.CustomTags, optionKeyPrefix, out optionValue))
+                {
+
```
```suggestion
```

Evangelink commented on src/Utilities/Compiler/Options/AbstractCategorizedAnalyzerConfigOptions.cs:
```diff
@@ -88,13 +89,23 @@ public bool TryGetOptionValue<T>(string optionName, OptionKind kind, DiagnosticD
-                TryGetAnySpecificOptionValue(rule.CustomTags, optionKeyPrefix, out optionValue) ||
-                TryGetGeneralOptionValue(optionKeyPrefix, out optionValue))
+
+            T optionValue;
```
Out of curiosity, is it to reduce allocation that you declare the variable before rather than doing `out var`?

mavasani replied:
No, the var is again needed in a separate if check below, hence it must be scoped to the whole method.

Evangelink commented on src/Utilities/Compiler/Options/AbstractCategorizedAnalyzerConfigOptions.cs:
```diff
@@ -88,13 +89,23 @@ public bool TryGetOptionValue<T>(string optionName, OptionKind kind, DiagnosticD
+            T optionValue;
+            if (rule != null)
+            {
+                if (TryGetSpecificOptionValue(rule.Id, optionKeyPrefix, out optionValue) ||
```
[suggestion] could be merged with outer if.

mavasani replied:
That required adding parentheses around the second checks and was making the code clumsy :-)

Evangelink commented on src/Utilities/Compiler/Extensions/CompilationExtensions.cs:
```diff
@@ -59,5 +61,15 @@ internal static bool TargetsDotNetFramework(this Compilation compilation)
+        /// </summary>
+        internal static bool IsWebProject(this Compilation compilation, AnalyzerOptions options, CancellationToken cancellationToken)
+        {
+            var usingMicrosoftNETSdkWeb = options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.UsingMicrosoftNETSdkWeb, compilation, cancellationToken);
```
Is this property valid only for aspnet core or also regular aspnet? If it's valid only for aspnet core, do we really want to get rid of the old exclusions? That would mean raising FPs for "old" projects.

mavasani replied:
It will be valid for all SDK based aspnet projects. @pranavkm Is there any MSBuild property that is set for regular aspnet projects that we can use?

pranavkm replied:
> Is there any MSBuild property that is set for non-SDK based aspnet projects that we can use?

Not that I know of.

mavasani replied:
I looked around and the only way I could identify is by looking at `ProjectTypeGuids` property.

```csharp
public const string WebAppProjectGuidString = "349C5851-65DF-11DA-9384-00065B846F21";
public const string WebSiteProjectGuidString = "E24C65DC-7377-472B-9ABA-BC803B73C61A";
```

I'll can add these property value checks unless we find a better way.

Evangelink commented:
Without forcing you to add too many tests, it might be nice to add a couple of tests for some of the cases mentioned here https://github.com/dotnet/roslyn-analyzers/issues/3835#issuecomment-657591877

mavasani commented:
> Without forcing you to add too many tests, it might be nice to add a couple of tests for some of the cases mentioned here

I am not sure if that provides any valuable test coverage addition. The analyzer itself is not detecting any special patterns, except for bailing out on externally visible types when a certain option is provided - basically if you set that option manually in .editorconfig, it will skip externally visible types for that project. It is not checking for any semantics around the symbols or compilation, so there will not be real value add.

Evangelink commented:
It just provides a minimal safety-net in case we change something in the rule later on but I agree value is minimal here. It could be valuable to have some "integration" test though (but that's not in the scope of this PR).

mavasani commented:
>  It could be valuable to have some "integration" test though (but that's not in the scope of this PR).

Yes, that would be valuable, and agree out of the scope of this PR.

mavasani commented:
@Evangelink I have added support for non-SDK projects and also added support for `api_surface` option for the analyzer. I have left the attribute checks as I am not sure if they are intended only for public API surface or not.

mavasani commented:
@Evangelink would you mind reviewing the latest commit when you get time? Thanks!
